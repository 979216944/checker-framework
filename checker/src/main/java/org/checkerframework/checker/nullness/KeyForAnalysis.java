package org.checkerframework.checker.nullness;

import java.util.List;
import java.util.Set;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.TypeMirror;
import org.checkerframework.common.basetype.BaseTypeChecker;
import org.checkerframework.framework.flow.CFAbstractAnalysis;
import org.checkerframework.framework.flow.CFAbstractValue;
import org.checkerframework.javacutil.Pair;

/** Boiler plate code to glue together all the parts the KeyFor dataflow classes. */
public class KeyForAnalysis extends CFAbstractAnalysis<KeyForValue, KeyForStore, KeyForTransfer> {
    // fields inidicates whether nullness_lite is enabled with the following features
    protected boolean NO_ALIASING;
    protected boolean ALL_KEY_EXIST;

    public KeyForAnalysis(
            BaseTypeChecker checker,
            KeyForAnnotatedTypeFactory factory,
            List<Pair<VariableElement, KeyForValue>> fieldValues,
            int maxCountBeforeWidening) {
        super(checker, factory, fieldValues, maxCountBeforeWidening);
        setNullnessLite();
    }

    public KeyForAnalysis(
            BaseTypeChecker checker,
            KeyForAnnotatedTypeFactory factory,
            List<Pair<VariableElement, KeyForValue>> fieldValues) {
        super(checker, factory, fieldValues);
        setNullnessLite();
    }

    @Override
    public KeyForStore createEmptyStore(boolean sequentialSemantics) {
        return new KeyForStore(this, sequentialSemantics);
    }

    @Override
    public KeyForStore createCopiedStore(KeyForStore store) {
        return new KeyForStore(store);
    }

    @Override
    public KeyForValue createAbstractValue(
            Set<AnnotationMirror> annotations, TypeMirror underlyingType) {

        if (!CFAbstractValue.validateSet(annotations, underlyingType, qualifierHierarchy)) {
            return null;
        }
        return new KeyForValue(this, annotations, underlyingType);
    }

    private void setNullnessLite() {
        NO_ALIASING =
                checker.hasOption("NullnessLite")
                        && (checker.getOption("NullnessLite") == null
                                || checker.getOption("NullnessLite")
                                        .toLowerCase()
                                        .contains("inva"));
        ALL_KEY_EXIST =
                checker.hasOption("NullnessLite")
                        && (checker.getOption("NullnessLite") == null
                                || checker.getOption("NullnessLite")
                                        .toLowerCase()
                                        .contains("mapk"));
    }
}
