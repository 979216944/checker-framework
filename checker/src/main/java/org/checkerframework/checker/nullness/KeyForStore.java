package org.checkerframework.checker.nullness;

import org.checkerframework.dataflow.analysis.FlowExpressions;
import org.checkerframework.framework.flow.CFAbstractAnalysis;
import org.checkerframework.framework.flow.CFAbstractStore;

public class KeyForStore extends CFAbstractStore<KeyForValue, KeyForStore> {
    public KeyForStore(
            CFAbstractAnalysis<KeyForValue, KeyForStore, ?> analysis, boolean sequentialSemantics) {
        super(analysis, sequentialSemantics);
    }

    protected KeyForStore(CFAbstractStore<KeyForValue, KeyForStore> other) {
        super(other);
    }

    /**
     * [Invalidation of Dataflow] When see -ANullnessLite or -ANullnessLite=inva, canAlias returns
     * false to disallow aliasing; Note KeyForSubChecker doesn't inherit from NullnessChecke; Thus,
     * changing KeyForStore is the only way to change the control flow.
     */
    @Override
    public boolean canAlias(FlowExpressions.Receiver a, FlowExpressions.Receiver b) {
        if (((KeyForAnalysis) analysis).NO_ALIASING) {
            return false;
        }
        return super.canAlias(a, b);
    }
}
