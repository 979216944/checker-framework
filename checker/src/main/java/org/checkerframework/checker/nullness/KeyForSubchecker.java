package org.checkerframework.checker.nullness;

import java.util.HashMap;
import java.util.Map;
import javax.annotation.processing.SupportedOptions;
import org.checkerframework.common.basetype.BaseTypeChecker;

/**
 * A type-checker for determining which values are keys for which maps. Typically used as part of
 * the compound checker for the nullness type system.
 *
 * @checker_framework.manual #map-key-checker Map Key Checker
 * @checker_framework.manual #nullness-checker Nullness Checker
 */
@SupportedOptions({"NullnessLite"})
public class KeyForSubchecker extends BaseTypeChecker {
    /**
     * NullnessLite option for Nullness Checker 1. [Initialization Checker] disabled as command-line
     * option. 2. [Invalidation of Dataflow] impure methods disallowed as command-line option. 3.
     * [Invalidation of Dataflow] aliasing disallowed, see method canAlias in NullnessStore.java
     */
    @Override
    public void initChecker() {
        super.initChecker();

        if (this.hasOption("NullnessLite")) {
            Map<String, String> nullness_lite = new HashMap<String, String>();
            nullness_lite.put("suppressWarnings", "uninitialized");
            nullness_lite.put("assumeSideEffectFree", null);
            this.addOptions(nullness_lite);
        }
    }
}
