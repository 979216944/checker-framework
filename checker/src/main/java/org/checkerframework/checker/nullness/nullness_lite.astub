// stub file for NullnessLite option
// assume all BoxedClass.valueOf() method are @Pure
// returned objects are equal by ==
import org.checkerframework.dataflow.qual.Pure;

package java.lang;

public final class Character {
    @Pure public static Character valueOf(char ch);
}

public final class Double {
    @Pure public static Double valueOf(double a1);
}

public final class Integer {
    @Pure public static Integer valueOf(int a1);
}

public final class Float {
    @Pure public static Float valueOf(float a1);
}

public final class Long {
    @Pure  public static Long valueOf(long a1);
}

public final class Short {
    @Pure public static Short valueOf(short a1);
}

public final class Byte {
    @Pure public static Byte valueOf(byte a1);
}

public final class Boolean {
    @Pure public static Boolean valueOf(boolean a1);
}
