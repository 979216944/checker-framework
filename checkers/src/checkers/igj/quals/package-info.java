/**
 * Contains qualifiers for type annotations used by the IGJ checker.
 *
 * @see checkers.igj.IGJChecker
 *
 * @manual #igj-checker IGJ Checker
 */
package checkers.igj.quals;
