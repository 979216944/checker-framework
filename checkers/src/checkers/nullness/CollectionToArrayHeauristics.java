package checkers.nullness;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeKind;
import javax.lang.model.util.ElementFilter;

import checkers.types.*;
import checkers.types.AnnotatedTypeMirror.*;
import checkers.util.TreeUtils;

import com.sun.source.tree.MethodInvocationTree;
import com.sun.source.tree.Tree;

/**
 * Handles calls to {@link java.util.Collection#toArray()} and determines
 * the appropriate nullness type of the returned value.
 */
public class CollectionToArrayHeauristics {
    private final ProcessingEnvironment env;
    private final NullnessAnnotatedTypeFactory factory;
    private final AnnotatedTypes atypes;

    private final ExecutableElement collectionToArrayObject;
    private final ExecutableElement collectionToArrayE;
    private final AnnotatedDeclaredType collectionType;

    public CollectionToArrayHeauristics(ProcessingEnvironment env,
            NullnessAnnotatedTypeFactory factory) {
        this.env = env;
        this.factory = factory;
        this.atypes = new AnnotatedTypes(env, factory);

        this.collectionToArrayObject = getMethod("java.util.Collection", "toArray", 0);
        this.collectionToArrayE = getMethod("java.util.Collection", "toArray", 1);
        this.collectionType = factory.fromElement(env.getElementUtils().getTypeElement("java.util.Collection"));
    }

    public void handle(MethodInvocationTree tree, AnnotatedExecutableType method) {
        if (isMethod(tree, collectionToArrayObject)
                || isMethod(tree, collectionToArrayE)) {
            boolean receiver = isNonNullReceiver(tree);
            boolean argument = isNonNullArgument(tree);

            setComponentNullness(receiver && argument, method.getReturnType());

            // TODO: we need a mechanism to prevent nullable collections
            // from inserting null elements into a nonnull arrays
            if (!receiver && argument && !tree.getArguments().isEmpty())
                setComponentNullness(receiver, method.getParameterTypes().get(0));
        }
    }

    private void setComponentNullness(boolean isNonNull, AnnotatedTypeMirror type) {
        assert type.getKind() == TypeKind.ARRAY;
        AnnotatedTypeMirror compType = ((AnnotatedArrayType)type).getComponentType();
        compType.clearAnnotations();
        compType.addAnnotation(isNonNull ? factory.NONNULL : factory.NULLABLE);
    }

    private boolean isNonNullReceiver(MethodInvocationTree tree) {
        // check receiver
        AnnotatedTypeMirror receiver = factory.getReceiver(tree);
        AnnotatedDeclaredType collection = (AnnotatedDeclaredType)atypes.asSuper(receiver, collectionType);
        assert collection != null;

        if (collection.getTypeArguments().isEmpty()
            || !collection.getTypeArguments().get(0).hasAnnotation(factory.NONNULL))
            return false;
        return true;
    }

    private boolean isNonNullArgument(MethodInvocationTree tree) {
        // check based on argument
        if (isMethod(tree, collectionToArrayE)) {
            assert !tree.getArguments().isEmpty() : tree;
            AnnotatedTypeMirror type = factory.getAnnotatedType(tree.getArguments().get(0));
            assert type.getKind() == TypeKind.ARRAY;
            AnnotatedArrayType arType = (AnnotatedArrayType)type;
            return arType.getComponentType().hasAnnotation(factory.NONNULL);
        }

        return true;
    }

    // TODO: duplicated code from MapGetHeauristics
    private boolean isMethod(Tree tree, ExecutableElement method) {
        if (!(tree instanceof MethodInvocationTree))
            return false;
        MethodInvocationTree methInvok = (MethodInvocationTree)tree;
        ExecutableElement invoked = TreeUtils.elementFromUse(methInvok);
        return isMethod(invoked, method);
    }

    private boolean isMethod(ExecutableElement questioned, ExecutableElement method) {
        return (questioned.equals(method)
                || env.getElementUtils().overrides(questioned, method,
                        (TypeElement)questioned.getEnclosingElement()));
    }

    private ExecutableElement getMethod(String typeName, String methodName, int params) {
        TypeElement mapElt = env.getElementUtils().getTypeElement(typeName);
        for (ExecutableElement exec : ElementFilter.methodsIn(mapElt.getEnclosedElements())) {
            if (exec.getSimpleName().contentEquals(methodName)
                    && exec.getParameters().size() == params)
                return exec;
        }
        throw new RuntimeException("Shouldn't be here!");
    }

}
