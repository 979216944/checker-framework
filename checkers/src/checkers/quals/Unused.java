package checkers.quals;

import static java.lang.annotation.ElementType.*;

import java.lang.annotation.*;

/**
 * Declares that the field may not be accessed if the receiver is of the
 * specified qualifier type.
 *
 * The checker that type-checks the {@code when} element value qualifier
 * verifies that property.
 *
 * <p><b>Example</b>
 * Consider a class, {@code Table}, with a locking field, {@code lock}.  The
 * lock is used when a {@code Table} instance is shared across threads.  When
 * running in a local thread, the {@code lock} field ought not to be used.
 *
 * You can declare this behavior in the following way:
 *
 * <pre><code>
 * class Table {
 *   private @Unused(when=LocalToThread.class) final Lock lock;
 *   ...
 * }
 * </code></pre>
 *
 * The appropriate checker (which type-checks the specified qualifier) is to
 * issue an error whenever the lock is used in a {@code LocalToThread}, like
 * in the following case:
 *
 * <pre><code>
 *   final @LocalToThread Table table = ....;
 *   table.lock.lock();
 * </code></pre>
 *
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({FIELD})
public @interface Unused {
    Class<? extends Annotation> when();
}
