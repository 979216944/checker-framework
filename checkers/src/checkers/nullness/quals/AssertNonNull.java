package checkers.nullness.quals;

import java.lang.annotation.*;

import checkers.nullness.NullnessChecker;

/**
 * Indicates that the method throws a {@link Throwable} if any of its
 * parameters is a {@code null}.
 *
 * @see NonNull
 * @see NullnessChecker
 * @manual #nullness-checker Nullness Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface AssertNonNull {
}
