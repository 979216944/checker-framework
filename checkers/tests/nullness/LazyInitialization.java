import checkers.nullness.quals.*;

public class LazyInitialization {
    @Nullable Object nullable;
    @NonNull  Object nonnull;
    @LazyNonNull Object lazy;
    @LazyNonNull Object lazy2 = null;

    void randomMethod() { }

    void testAssignment() {
        lazy = "m";
        lazy = null;    // null
    }

    void testLazyBeingNull() {
        nullable.toString(); // error
        nonnull.toString();
        lazy.toString();    // error
    }

    void testAfterInvocation() {
        nullable = "m";
        nonnull = "m";
        lazy = "m";

        randomMethod();

        nullable.toString();    // error
        nonnull.toString();
        lazy.toString();
    }
}
