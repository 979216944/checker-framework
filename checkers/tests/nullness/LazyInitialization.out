:13: (type.incompatible)
:17: (dereference.of.nullable)
:19: (dereference.of.nullable)
:29: (dereference.of.nullable)
